File: python/infinicore/ops/trmm.py
from infinicore.lib import _infinicore
from infinicore.tensor import Tensor


def trmm(
    a: Tensor,
    alpha: Tensor,
    b: Tensor,
    *,
    side: int = 0,
    uplo: int = 0,
    trans: int = 0,
    diag: int = 0,
):
    _infinicore.trmm_(
        a._underlying,
        alpha._underlying,
        b._underlying,
        side,
        uplo,
        trans,
        diag,
    )

    return b

File: src/infinicore/ops/trmm/trmm.cc
#include "infinicore/ops/trmm.hpp"

#include "../../utils.hpp"

namespace infinicore::op {

INFINICORE_GRAPH_OP_DISPATCHERS_IMPL(Trmm);

Trmm::Trmm(const Tensor &a, const Tensor &alpha, Tensor b, int side, int uplo, int trans, int diag) {
    INFINICORE_ASSERT_TENSORS_SAME_DEVICE(a, alpha, b);
    INFINICORE_GRAPH_OP_DISPATCH(b->device().getType(), a, alpha, b, side, uplo, trans, diag);
}

void Trmm::execute(const Tensor &a, const Tensor &alpha, Tensor b, int side, int uplo, int trans, int diag) {
    INFINICORE_GRAPH_OP_RECORD_OR_RUN(Trmm, a, alpha, b, side, uplo, trans, diag);
}

void trmm_(const Tensor &a, const Tensor &alpha, Tensor b, int side, int uplo, int trans, int diag) {
    Trmm::execute(a, alpha, b, side, uplo, trans, diag);
}

} // namespace infinicore::op
